### PR TITLE
Position e2e Mount Propagation test to be kubelet-specific

### LINKED ISSUE #100259: Mount Propagation to the Host should not be unneeded for k8s e2e tests
#### What would you like to be changed:

The `Mount propagation` e2e test should be relaxed.  It currently checks 3 different aspects of the `mountPropagation` volume mount flag:

1. Containers with `mountPropagation: default` do not receive mounts from outside their own mount namespace, nor do mount namespaces outside the container receive mounts created within the container.
2. Containers with `mountPropagation: hostToContainer` receive new mounts from the host OS, the container runtime, and other containers with `mountPropagation: Bidirectional`, but no mount namespaces outside the container receive mounts created within the container.
3. Containers with `mountPropagation: Bidirectional` receive new mounts identical to containers with `mountPropagation: hostToContainer`, but additionally any mounts made by these containers are:
    1. propagated to the container runtime where they can be propagated to other containers (related to _2_)
    2. also propagated to the Host OS

The first 3 aspects, up to and including _3.i_, are critical for container operation and should still be ensured by an e2e test.  This last check, _3.ii._, is the part that I believe is overly strict.

#### Why is this needed:

The current e2e test is too opinionated on the underlying node OS configuration. The added requirement of propagation to the host OS precludes administrators or solution providers from modifying the Kubernetes installation to hide these Kubernetes-specific mountpoints, which can be for a variety of reasons:

##### Security

While access to container-specific mounts in Kubernetes (like secrets) today are protected by filesystem permissions, the fact that the locations are visible to unprivileged users via the 'mount' command can provide information which could be used in conjunction with other vulnerabilities to decipher secrets or other protected container mount contents. Locking access to even
listing the mountpoints behind the ability to enter the mount namespace (which in turn requires `CAP_SYS_CHROOT` and `CAP_SYS_ADMIN`) removes one level of potential access to container-specific mounts.

##### Tidiness and Encapsulation

These container overlay mounts, secret and configmap mounts, and namespace mounts are all internal implementation details around the way the Kubernetes and the Container Runtime allocate resources to containers. For normal day-to-day operation, there is no need for these special-use mounts to be visible in the host OS's top-level 'mount' or 'findmnt' output.

##### No one uses it (that we know of)

After much investigation and conversation on both the sig-node and sig-storage mailing lists, as well as within Red Hat's OpenShift communities, we have yet to turn up a use case that relies on this container-to-host propagation.

##### No one should use it

In fact, it may be seen as an anti-pattern and run contrary to the container-native way of doing things.  Relying on a host OS to interact with a container-owned mountpoint seems suspect.

##### Host OS Efficiency

Today systemd's CPU usage is adversely affected by a large number of mountpoints, especially versions released prior to December 2020. This feature of systemd, which scans all /proc/1/mount to invoke udev events for every change suffers from two issues. The first was an inefficiency in systemd’s processing logic. This was fixed in December 2020, and this fix reportedly reduced the systemd mount-scanning-related overhead by 30-50%. The second is a lack of granularity in the events the kernel sends to systemd about mountpoint changes, which necessitates a full rescan of all mountpoints even on a minor change or single mountpoint addition. The kernel change to remedy this seems unlikely to move forward in the near term. (See [BZ1819869](https://bugzilla.redhat.com/show_bug.cgi?id=1819868) for more details)  

For example, a fairly small OpenShift deployment today (which does not yet have the systemd December 2020 fix) has upwards of 400 mountpoints on every node owned by Kubelet, the Container Runtime, and infrastructure containers like OVN or CSI operators. Hiding the container mountpoints reduces the steady-state systemd CPU utilization from ~10% to ~0% on every worker node.  On a more heavily loaded system, with 1000’s of container mountpoints, systemd’s usage will cap out at 100% of a single CPU core.  Even with the 30-50% reduction in systemd usage expected from the December 2020 fix, having systemd take up as much as 50-70% of a single CPU in the worst-case scenario is a lot of unnecessary overhead.

Administrators of other OSes with other host OS services that monitor or process mountpoints indiscriminately would be able to hide these container-specific mountpoints from the OS at large, at their option.

##### Testing implications

With this requirement relaxed, it would be possible to run multiple instances of Kubernetes on a single host without the various mountpoints interfering with one another.  There may be other inter-instance difficulties besides the mount namespace, but this does move in that direction without resorting to more complicated solutions like trying to run Kubernetes inside of containers.

### PR DESCRIPTION
**Pull request by lack**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This replaces the check of mount propagation to/from the host OS mount namespace to a similar check about the mount namespace where kubelet is running (which may or may not be the same mount namespace as the host OS).

With this change in place, I believe this is now a good candidate for future promotion to a conformance test.

#### Which issue(s) this PR fixes:

Fixes #100259

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/node/mount_propagation.go [hunk 1]
```diff
@@ -180,15 +181,28 @@ var _ = SIGDescribe("Mount propagation", func() {
...
-		cmd = fmt.Sprintf("test `cat %q/master/file` = master", hostDir)
-		err = hostExec.IssueCommand(cmd, node)
-		framework.ExpectNoError(err, "host should see mount from master")
 
-		// Host can't see mount from slave
-		cmd = fmt.Sprintf("test ! -e %q/slave/file", hostDir)
-		err = hostExec.IssueCommand(cmd, node)
-		framework.ExpectNoError(err, "host shouldn't see mount from slave")
+		// Find the kubelet PID to ensure we're working with the kubelet's mount namespace
+		cmd = "pidof /usr/bin/kubelet"
```
Comment (derekwaynecarr) [inline test/e2e/node/mount_propagation.go:186]: per discussion in sig, we will want to use an alternate mechanism to discover the kubelet pid.  we may have a helper elsewhere, will have to look.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]

File: test/e2e/node/mount_propagation.go (hunk 1 above)
Comment (dims) [inline test/e2e/node/mount_propagation.go:186]: we need to find a better way than using the full path of the kubelet as this is different across distros

Comment (lack) [reply test/e2e/node/mount_propagation.go:186]: I suppose I could use `pidof kubelet` and hope there isn't another process with the same name...

Comment (lack) [reply test/e2e/node/mount_propagation.go:186]: Agreed.

Comment (lack) [reply test/e2e/node/mount_propagation.go]: I looked elsewhere in the code, and it looks like the convention in other e2e tests here is to just use `pidof execname` without an explicit path.

**Comment by lack** [comment]
@derekwaynecarr and @dims : Does this implementation look better to you?  I'm now calling `pidof kubelet` instead of `pidof /usr/bin/kubelet` as I originally had.

**Comment by dims** [comment]
@lack as long as it consistently works across OS's :) +1

**Comment by derekwaynecarr** [comment]
/ok-to-test

there should only be one pidof kubelet in the context this is invoked, so i think we are good.

**Comment by derekwaynecarr** [comment]
will tag when we get clean ci signal.

**Comment by derekwaynecarr** [comment]
failure was in unrelated test.

/test pull-kubernetes-node-e2e

**Comment by lack** [comment]
@derekwaynecarr : CI is happy!

**Comment by lack** [comment]
This test doesn't have a Details link - is it stuck?
/test pull-kubernetes-node-e2e-containerd

**Review by derekwaynecarr — APPROVED** [review]

**Pull request merged**